notes/2020-10-02.md

```diff
@@ -30,52 +30,58 @@ ca_dds_df
 ```{code-cell} ipython3
 print(ca_dds_df)
 ```
-
+we can do df.index to look at what the index is in our dataframe
 ```{code-cell} ipython3
 ca_dds_df.index
 ```
-
+columns are tuple values. We can index to the first column as below. As we can see by running this code, this column has three names initially. 
 ```{code-cell} ipython3
 ca_dds_df.columns[0]
 ```
-
+By the following code we can assign the first column as the index of original dataframe. If we assign "False" to the "inplace", it will make the changes in a new dataframe.
 ```{code-cell} ipython3
 ca_dds_df.set_index(ca_dds_df.columns[0],inplace=True)
 ```
-
+The updated dataframe looks like below. We still have the name of index inappropriate and aslo there are empty rows. 
 ```{code-cell} ipython3
 ca_dds_df
 ```
-
+we can drop rows with "all" or "any",  "NaN" values. We would choose "all", since if we choose "any", we will loose many rows here and we do not want it.
 ```{code-cell} ipython3
 ca_dds_df.dropna(how='all',inplace=True)
 ```
-
+We can view the index by the following code. Remember that this is not a regular column anymore since we assigned it to be our dataframe index.
 ```{code-cell} ipython3
 ca_dds_df.index
 ```
-
+Now lets rename the index to a more meaningful name as below.
 ```{code-cell} ipython3
 ca_dds_df.index.rename('Age Cohort',inplace=True)
 ```
+But we still have multi-row headers which should get corrected. We also want to have columns for "Age", "Race", "Gender", "Mean" and "Count".
 
+
+
+Let's flip the dataframe and unstack the index. For this purpose, we should unstack it in two levels of "Race" and "Gender".  After unstacking we want to flip it again and finally reset the index.
 ```{code-cell} ipython3
 ca_dds_df.T.unstack(level= [0,1]).T.reset_index()
 ```
-
+All these changes will be saved into a new dataframe named "ca_dds_clean".
 ```{code-cell} ipython3
 ca_dds_clean = ca_dds_df.T.unstack(level= [0,1]).T.reset_index()
 ca_dds_clean.head()
 ```
-
+Now let's change name of columns to appropriate names. 
+As we can see here, we can not change elements of an index. Instead we should reassign the whole column names.
+The reason is that index variable's type is type "index" not a " list", "tuple" or a "regular python based datatype". Index has special properties and python does not allow mutaple operations or assigning values to that. 
 ```{code-cell} ipython3
 ca_dds_clean.columns[1] = 'Race'
 ```
```

Comment:
```suggestion

Now let's change name of columns to be formatted better for Python. We want to rename the one in index 1 to race. 
```{code-cell} ipython3
ca_dds_clean.columns[1] = 'Race'
```
```suggestion
As we can see here, we can not change elements of an index. Instead we should reassign the whole column names.
The reason is that index variable's type is type `pandas` `Index` not a regular python based datatype.  Index has special properties and python does not allow mutable operations, such as assigning it a new value. 
```

```